### DISCUSSION TITLE
How to read back data from a rp2040 flash storage with littlefs

### DISCUSSION THREAD

**Discussion by draeron** [tinygo-org/tinygo#4281] [Q&A]
I have a rp2040 microcontroller with 2MB of flash storage (waveshare 1.28). I plan to setup a NFC reader with it and log scans into the FS with littleFS. Anyone know how it would be possible to read "back" the data collected from the controller with the integrated USB?

**Comment by deadprogram** [comment]
This will require implementing the USB Mass Storage Device (MSD) side in TinyGo. Which is something that I also very much want myself.

**Reply by mikesmitty** [reply]
This is something that is relevant to my interests as well. I saw [this example](https://github.com/oyama/pico-littlefs-usb) emulating FAT12 on a pico today and was thinking about porting it to Go. Would this make sense to be included in the TinyFS repo?

Edit: after digging into it further it's using tinyusb for the USB msc portion. Porting that too is probably a bit more than I'm willing to take on at the moment

**Comment by Gustavomurta** [comment]
It's not clear to me. You can write the data to a mini SD card attached to the RP2040. Or do you want to write the data to a USB flash drive?

**Reply by draeron** [reply]
i want to be read back the data from the littleFS flash storage

**Reply by Gustavomurta** [reply]
Unfortunately Tinygo does not support USB flash drive access.

**Reply by sago35** [reply]
Currently, TinyGo does not implement USB MSD, so there is no way to mount and read directly. The options might be as follows:

* Implement USB MSD
* Implement a custom rule/protocol over USB CDC to retrieve the data

An example of tinyfs + USB CDC can be found below.

https://github.com/tinygo-org/tinyfs/blob/v0.4.0/examples/console/example.go#L477-L515
